### DISCUSSION TITLE
Why iperf can't reach higher bandwidth

### DISCUSSION THREAD

**Discussion by ZENOTME** [shadow/shadow#3408] [Q&A]
Hi, I have the following config and hope to simulate high bandwidth(1000Gbit) iperf.
```
general:
  stop_time: 200s
experimental:
  scheduler: thread-per-host
network:
  graph:
    type: gml
    inline: |
      graph [
        directed 0
        node [
          id 0
          host_bandwidth_up "1000 Gbit"
          host_bandwidth_down "1000 Gbit"
        ]
        edge [
          source 0
          target 0
          latency "1 ns"
          packet_loss 0.0
        ]
      ]
hosts:
  server:
    network_node_id: 0
    processes:
    - path: iperf
      args: -s
      start_time: 0s
      expected_final_state: running
  client:
    network_node_id: 0
    processes:
    - path: iperf
      args: -c server -t 20
      start_time: 2s
```

But the actual result(5.73 Gbits/sec) is smaller than this expected. I'm not familiar with DES, in theory, it can take more time to simulate a higher bandwidth but why it can't? Did I config wrong or are there other things that limit this?

**Comment by robgjansen** [comment]
I'm not sure why you need a terabit/s switch, but two things are likely happening here: latency and packet loss:

- latency: you set a latency of 1 nanosecond, but shadow will probably adjust that to  1 millisecond with its runahead feature (https://shadow.github.io/docs/guide/shadow_config_spec.html#experimentalrunahead)
- packet loss: Shadow's simulated routing layer uses an implementation of CoDel, which will cause packet loss even if you configure 0 packet loss on your network graph.
 
Packet loss and latency, in addition to protocol overhead, will limit the throughput that can be achieved with TCP.

**Reply by ZENOTME** [reply]
> I'm not sure why you need a terabit/s switch, but two things are likely happening here: latency and packet loss:
> 
> * latency: you set a latency of 1 nanosecond, but shadow will probably adjust that to  1 millisecond with its runahead feature (https://shadow.github.io/docs/guide/shadow_config_spec.html#experimentalrunahead)
> * packet loss: Shadow's simulated routing layer uses an implementation of CoDel, which will cause packet loss even if you configure 0 packet loss on your network graph.
> 
> Packet loss and latency, in addition to protocol overhead, will limit the throughput that can be achieved with TCP.

Thanks for your reply! I try to set runahead to null and the throughput increased to 8Gbit but still far less than expected. I keep the same TCP setting, e.g. the send buffer and receive buffer in the real case, and find that we can achieve a higher throughput in the real case. 

> Packet loss and latency, in addition to protocol overhead, will limit the throughput that can be achieved with TCP.

So does this mean that some simulations are inaccurate for this scenario? Are there other models I can change to bypass some simulation e.g. TCP or non-packet loss?

 > I'm not sure why you need a terabit/s switch

I want to test how much time this simulator should take for high throughput such as terabit.

**Reply by robgjansen** [reply]
> > Packet loss and latency, in addition to protocol overhead, will limit the throughput that can be achieved with TCP.

> So does this mean that some simulations are inaccurate for this scenario? Are there other models I can change to bypass some simulation e.g. TCP or non-packet loss?

We have started some work on reimplementing the Shadow TCP stack in rust. It is currently an experimental feature because it doesn't support important aspects of TCP, but you could try it out using the [use_new_tcp](https://shadow.github.io/docs/guide/shadow_config_spec.html#experimentaluse_new_tcp) flag and see if that can reach higher throughput.

> I keep the same TCP setting, e.g. the send buffer and receive buffer in the real case, and find that we can achieve a higher throughput in the real case.

Could you explain what is your real case here? What is your switching capacity, and what is the real throughput you achieve with TCP?
